lib/codegen/CreateTerm.cpp

```diff
@@ -0,0 +1,198 @@
+
+llvm::Value *getBlockHeader(llvm::Module *Module, const KOREObjectSymbol *symbol) {
+  llvm::StructType *BlockHeaderType = Module->getTypeByName(BLOCKHEADER_STRUCT);
+  uint64_t headerVal = symbol->getTag();
+  headerVal |= symbol->length() << 32;
+  headerVal |= symbol->getLayout() << 48;
+  return llvm::ConstantStruct::get(BlockHeaderType, llvm::ConstantInt::get(llvm::IntegerType::get(Module->getContext(), 64), headerVal));
+}
+
+llvm::Value *allocateBlock(const KOREObjectSymbol *symbol, llvm::Type *AllocType, llvm::BasicBlock *block) {
+  llvm::Instruction *Malloc = llvm::CallInst::CreateMalloc(block, llvm::IntegerType::get(block->getContext(), 64), AllocType, llvm::ConstantExpr::getSizeOf(AllocType), nullptr, nullptr);
+  block->getInstList().push_back(Malloc);
+  return Malloc;
+}
+
+llvm::Type *termType(KOREPattern *pattern, llvm::StringMap<llvm::Value *> &substitution, KOREDefinition *definition, llvm::LLVMContext &Context, llvm::Module *Module) {
+  if (auto variable = dynamic_cast<KOREObjectVariablePattern *>(pattern)) {
+    return substitution.lookup(variable->getName())->getType();
+  } else if (auto constructor = dynamic_cast<KOREObjectCompositePattern *>(pattern)) {
+    const KOREObjectSymbol *symbol = constructor->getConstructor();
+    assert(symbol->isConcrete() && "not supported yet: sort variables");
+    assert(symbol->getName() != "\\dv" && "not supported yet: \\dv");
+    KOREObjectSymbolDeclaration *symbolDecl = definition->getSymbolDeclarations().lookup(symbol->getName());
+    if (symbolDecl->getAttributes().count("function")) {
+      assert(false && "not implemented yet: functions");
+    }
+    return llvm::PointerType::get(Module->getTypeByName(BLOCK_STRUCT), 0);
+  } else {
+    assert(false && "not supported yet: meta level");
+  }
+}
+  
+
+llvm::Value *createTerm(KOREPattern *pattern, llvm::StringMap<llvm::Value *> &substitution, KOREDefinition *definition, llvm::BasicBlock *block, llvm::Module *Module) {
+  if (auto variable = dynamic_cast<KOREObjectVariablePattern *>(pattern)) {
+    return substitution.lookup(variable->getName());
+  } else if (auto constructor = dynamic_cast<KOREObjectCompositePattern *>(pattern)) {
+    const KOREObjectSymbol *symbol = constructor->getConstructor();
+    assert(symbol->isConcrete() && "not supported yet: sort variables");
+    assert(symbol->getName() != "\\dv" && "not supported yet: \\dv");
+    KOREObjectSymbolDeclaration *symbolDecl = definition->getSymbolDeclarations().lookup(symbol->getName());
+    llvm::LLVMContext &Context = block->getContext();
+    if (symbolDecl->getAttributes().count("function")) {
+      assert(false && "not implemented yet: functions");
+    } else if (symbol->getArguments().empty()) {
+      llvm::StructType *BlockType = Module->getTypeByName(BLOCK_STRUCT);
+      llvm::IntToPtrInst *Cast = new llvm::IntToPtrInst(llvm::ConstantInt::get(llvm::IntegerType::get(Context, 32), symbol->getTag()), llvm::PointerType::get(BlockType, 0), "", block);
```

Comment:
This does not ensure the least bit is set. You said the tag would be put in the high 32 bits, and and a 32-bit 1 in the low half.